libevmasm/Assembly.cpp

```diff
@@ -884,175 +918,292 @@ LinkerObject const& Assembly::assemble() const
+				m_tagPositionsInBytecode[0] = ret.bytecode.size();
+
+			switch (i.type())
 			{
-				if (i != offsets.size() - 1)
+				case Operation:
+					ret.bytecode.push_back(static_cast<uint8_t>(i.instruction()));
+					break;
+				case Push:
 				{
-					ret.bytecode.push_back(uint8_t(Instruction::DUP2));
-					ret.bytecode.push_back(uint8_t(Instruction::DUP2));
+					unsigned b = numberEncodingSize(i.data());
+					if (b == 0 && !m_evmVersion.hasPush0())
+					{
+						b = 1;
+					}
+					ret.bytecode.push_back(static_cast<uint8_t>(pushInstruction(b)));
+					if (b > 0)
+					{
+						ret.bytecode.resize(ret.bytecode.size() + b);
+						bytesRef byr(&ret.bytecode.back() + 1 - b, b);
+						toBigEndian(i.data(), byr);
+					}
+					break;
 				}
-				// TODO: should we make use of the constant optimizer methods for pushing the offsets?
-				bytes offsetBytes = toCompactBigEndian(u256(offsets[i]));
-				ret.bytecode.push_back(static_cast<uint8_t>(pushInstruction(static_cast<unsigned>(offsetBytes.size()))));
-				ret.bytecode += offsetBytes;
-				ret.bytecode.push_back(uint8_t(Instruction::ADD));
-				ret.bytecode.push_back(uint8_t(Instruction::MSTORE));
-			}
-			if (offsets.empty())
-			{
-				ret.bytecode.push_back(uint8_t(Instruction::POP));
-				ret.bytecode.push_back(uint8_t(Instruction::POP));
+				case PushTag:
+				{
+					assertThrow(!eof, AssemblyException, "PushTag in EOF code");
+					ret.bytecode.push_back(tagPush);
+					tagRef[ret.bytecode.size()] = i.splitForeignPushTag();
+					ret.bytecode.resize(ret.bytecode.size() + bytesPerTag);
+					break;
+				}
+				case PushData:
+					// assertThrow(!eof, AssemblyException, "Push data in EOF code");
+					ret.bytecode.push_back(dataRefPush);
+					dataRef.insert(std::make_pair(h256(i.data()), ret.bytecode.size()));
+					ret.bytecode.resize(ret.bytecode.size() + bytesPerDataRef);
+					break;
+				case PushSub:
+					assertThrow(!eof, AssemblyException, "PushSub in EOF code");
+					assertThrow(i.data() <= std::numeric_limits<size_t>::max(), AssemblyException, "");
+					ret.bytecode.push_back(dataRefPush);
+					subRef.insert(std::make_pair(static_cast<size_t>(i.data()), ret.bytecode.size()));
+					ret.bytecode.resize(ret.bytecode.size() + bytesPerDataRef);
+					break;
+				case PushSubSize:
+				{
+					assertThrow(!eof, AssemblyException, "PushSubSize in EOF code");
+					assertThrow(i.data() <= std::numeric_limits<size_t>::max(), AssemblyException, "");
+					auto s = subAssemblyById(static_cast<size_t>(i.data()))->assemble().bytecode.size();
+					i.setPushedValue(u256(s));
+					unsigned b = std::max<unsigned>(1, numberEncodingSize(s));
+					ret.bytecode.push_back(static_cast<uint8_t>(pushInstruction(b)));
+					ret.bytecode.resize(ret.bytecode.size() + b);
+					bytesRef byr(&ret.bytecode.back() + 1 - b, b);
+					toBigEndian(s, byr);
+					break;
+				}
+				case PushProgramSize:
+				{
+					assertThrow(!eof, AssemblyException, "Push program size in EOF code");
```

Comment:
I think the style you used for these messages above was much clearer. Here it sounds like it's telling me to push program size into the code which makes no sense :)

```suggestion
					assertThrow(!eof, AssemblyException, "PushProgramSize in EOF code");
```

Reply:
Also, why is it commented out for `PushData` above?